File: src/org/cerion/symcalc/expression/number/RationalNum.java
package org.cerion.symcalc.expression.number;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.NumberExpr;
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import java.math.BigDecimal;

public class RationalNum extends NumberExpr
{
	public static final RationalNum ZERO = new RationalNum(IntegerNum.ZERO, IntegerNum.ONE);
	public static final RationalNum ONE = new RationalNum(IntegerNum.ONE, IntegerNum.ONE);

	public RationalNum(IntegerNum n) {
		this(n, IntegerNum.ONE);
	}

	public RationalNum(IntegerNum n, IntegerNum d) {
		if(d.getSignum() == -1)
			set(n.unaryMinus(),d.unaryMinus());
		else
			set(n,d);
	}

	public RationalNum(int n, int d) {
		this(new IntegerNum(n), new IntegerNum(d));
	}

	public IntegerNum numerator() {
		return (IntegerNum)get(0);
	}
	
	public IntegerNum denominator() {
		return (IntegerNum)get(1);
	}

	@Override
	public boolean equals(NumberExpr e) {
		if(e.isRational()) {
			RationalNum r = (RationalNum) e;
			return numerator().equals(r.numerator()) && denominator().equals(r.denominator());
		}

		return false;
	}

	@Override
	public NumberExpr evaluate() {
		//Reduce with GCD
		IntegerNum gcd = numerator().gcd(denominator());
		IntegerNum n;
		IntegerNum d;

		if(!gcd.isOne()) {
			n = (IntegerNum) numerator().div(gcd);
			d = (IntegerNum) denominator().div(gcd);
			set(n,d);
		}

		//Never allow bottom to be negative
		if(denominator().getSignum() == -1) {
			n = numerator().unaryMinus();
			d = denominator().unaryMinus();
			set(n,d);
		}

		//Integer since denominator is one
		if(denominator().isOne()) {
			return numerator();
		}

		return this;
	}

	@Override
	public int numType() {
		return RATIONAL;
	}

	@Override
	public String toString() {
		return numerator().toString() + "/" + denominator().toString();
	}

	@Override
	public double toDouble() {
		BigDecimal decN = numerator().toBigDecimal();
		BigDecimal decD = denominator().toBigDecimal();
		decN = decN.divide(decD);
		return decN.doubleValue();
	}

	@Override
	public NumberExpr unaryMinus() {
		return new RationalNum(numerator().unaryMinus(), denominator());
	}

	public NumberExpr reciprocal() {
		return new RationalNum(denominator(), numerator()).evaluate();
	}

	@Override
	public boolean isZero() {
		return numerator().isZero();
	}

	@Override
	public boolean isOne() {
		return numerator().equals(denominator());
	}

	@Override
	public NumberExpr plus(NumberExpr num) {
		RationalNum result;
		switch (num.numType()) 
		{
			case INTEGER: //RationalNum + IntegerNum
				IntegerNum i = num.asInteger();

				IntegerNum norm = i.times(denominator());
				return new RationalNum(numerator().plus(norm), denominator());
				
			case RATIONAL: //RationalNum + RationalNum
			{
				RationalNum div = (RationalNum)num;
				IntegerNum n1 = div.numerator().times( denominator());
				IntegerNum n2 = numerator().times( div.denominator() );
				
				result = new RationalNum(n1.plus(n2), denominator().times(div.denominator()));
				return result.evaluate();
			}
		}
		
		return num.plus(this); //Default reverse order
	}

	@Override
	public NumberExpr minus(NumberExpr num) {
		NumberExpr negative = num.unaryMinus();
		return this.plus(negative);
	}

	@Override
	public NumberExpr times(NumberExpr num) {
		RationalNum result;
		switch (num.numType()) 
		{
			case INTEGER: //RationalNum * IntegerNum
			{
				result = new RationalNum(numerator().times(num.asInteger()), denominator());
				return result.evaluate();
			}
			case RATIONAL: //RationalNum * RationalNum
			{
				RationalNum t = (RationalNum)num;
				result = new RationalNum( numerator().times( t.numerator() ), denominator().times( t.denominator() ) );
				return result.evaluate();
			}
		}
		
		return num.times(this);
	}

	@Override
	public NumberExpr div(NumberExpr num) {
		RationalNum result;
		switch (num.numType()) 
		{
			case INTEGER: //RationalNum / IntegerNum
			{
				result = new RationalNum(numerator(), denominator().times( num.asInteger() ) );
				return result.evaluate();
			}
			case RATIONAL: //RationalNum / RationalNum
			{
				RationalNum t = (RationalNum)num;
				result = new RationalNum( numerator().times( t.denominator()), denominator().times(t.numerator()) );
				return result.evaluate();
			}
		}
		
		return num.times(this);
	}

	@Override
	public boolean canExp(NumberExpr num) {
		if(num.numType() == RATIONAL || num.numType() == COMPLEX)
			return false;

		return true;
	}

	@Override
	public Expr power(NumberExpr num) {
		IntegerNum n;
		IntegerNum d;
		switch (num.numType()) 
		{
			case INTEGER: //RationalNum ^ IntegerNum
				n = (IntegerNum) numerator().power( num );
				d = (IntegerNum) denominator().power( num );
				return new RationalNum(n,d);
			
			//case RATIONAL: //RationalNum ^ RationalNum
			//	break;
				
			case REAL: //RationalNum ^ RealNum
			{
				RealNum rResult = RealNum.create(this);
				return rResult.power(num);
			}
		}
	
		return null;
	}

	private void set(IntegerNum n, IntegerNum d) {
		if(n != null)
			setArg(0, n);
		if(d != null)
			setArg(1, d);
	}

	@Override
	public int compareTo(NumberExpr o) {
		throw new NotImplementedException();
	}
}

File: src/org/cerion/symcalc/expression/number/RationalNum.kt
package org.cerion.symcalc.expression.number

import org.cerion.symcalc.expression.ErrorExpr
import org.cerion.symcalc.expression.Expr

class RationalNum @JvmOverloads constructor(n: IntegerNum, d: IntegerNum = IntegerNum.ONE) : NumberExpr() {

    override val isZero: Boolean get() = numerator.isZero
    override val isOne: Boolean get() = numerator.equals(denominator)
    override val numType: NumberType get() = NumberType.RATIONAL

    val numerator: IntegerNum get() = get(0) as IntegerNum
    val denominator: IntegerNum get() = get(1) as IntegerNum

    init {
        if (d.signum == -1)
            set(n.unaryMinus(), d.unaryMinus())
        else
            set(n, d)
    }

    constructor(n: Int, d: Int) : this(IntegerNum(n.toLong()), IntegerNum(d.toLong()))

    // TODO add compareTo and equals can build off that
    override fun equals(e: NumberExpr): Boolean {
        if (e.isRational) {
            val r = e as RationalNum
            return numerator.equals(r.numerator) && denominator.equals(r.denominator)
        }

        return false
    }

    override fun evaluate(): NumberExpr {
        //Reduce with GCD
        val gcd = numerator.gcd(denominator)
        var n: IntegerNum
        var d: IntegerNum

        if (!gcd.isOne) {
            n = numerator.div(gcd).asInteger()
            d = denominator.div(gcd).asInteger()
            set(n, d)
        }

        //Never allow bottom to be negative
        if (denominator.signum == -1) {
            n = numerator.unaryMinus()
            d = denominator.unaryMinus()
            set(n, d)
        }

        //Integer since denominator is one
        return if (denominator.isOne) numerator else this
    }

    override fun toString(): String = numerator.toString() + "/" + denominator.toString()

    override fun toDouble(): Double {
        var decN = numerator.toBigDecimal()
        val decD = denominator.toBigDecimal()
        decN = decN.divide(decD)
        return decN.toDouble()
    }

    override fun unaryMinus(): NumberExpr = RationalNum(numerator.unaryMinus(), denominator)

    fun reciprocal(): NumberExpr = RationalNum(denominator, numerator).evaluate()

    override fun plus(number: NumberExpr): NumberExpr {
        val result: RationalNum
        when (number.numType) {
            NumberType.INTEGER -> {
                val norm = number.asInteger() * denominator
                return RationalNum(numerator.plus(norm), denominator)
            }

            NumberType.RATIONAL -> {
                val div = number as RationalNum
                val n1 = div.numerator * denominator
                val n2 = numerator * div.denominator

                result = RationalNum(n1 + n2, denominator * div.denominator)
                return result.evaluate()
            }
        }

        return number + this //Default reverse order
    }

    override fun minus(number: NumberExpr): NumberExpr = this + number.unaryMinus()

    override fun times(num: NumberExpr): NumberExpr {
        val result: RationalNum
        when (num.numType) {
            NumberType.INTEGER -> {
                result = RationalNum(numerator * num.asInteger(), denominator)
                return result.evaluate()
            }
            NumberType.RATIONAL -> {
                val t = num as RationalNum
                result = RationalNum(numerator * t.numerator, denominator * t.denominator)
                return result.evaluate()
            }
        }

        return num * this
    }

    override fun div(num: NumberExpr): NumberExpr {
        val result: RationalNum
        when (num.numType) {
            NumberType.INTEGER -> {
                result = RationalNum(numerator, denominator * num.asInteger())
                return result.evaluate()
            }
            NumberType.RATIONAL -> {
                val t = num as RationalNum
                result = RationalNum(numerator * t.denominator, denominator * t.numerator)
                return result.evaluate()
            }
        }

        return num.times(this)
    }

    override fun canExp(num: NumberExpr): Boolean = !(num.isRational || num.isComplex)

    override fun power(num: NumberExpr): Expr {
        val n: IntegerNum
        val d: IntegerNum
        when (num.numType) {
            NumberType.INTEGER -> {
                n = numerator.power(num) as IntegerNum
                d = denominator.power(num) as IntegerNum
                return RationalNum(n, d)
            }

            //case RATIONAL: //RationalNum ^ RationalNum
            //	break;

            NumberType.REAL -> {
                val rResult = RealNum.create(this)
                return rResult.power(num)
            }
        }

        return ErrorExpr("Not implemented")
    }

    private operator fun set(n: IntegerNum?, d: IntegerNum?) {
        if (n != null)
            setArg(0, n)
        if (d != null)
            setArg(1, d)
    }

    override fun compareTo(o: NumberExpr): Int {
        throw NotImplementedError()
    }

    companion object {
        val ZERO = RationalNum(IntegerNum.ZERO, IntegerNum.ONE)
        val ONE = RationalNum(IntegerNum.ONE, IntegerNum.ONE)
    }
}